src/test/java/org/breedinginsight/api/v1/controller/brapi/BrAPIServerFilterIntegrationTest.java

```diff
@@ -145,88 +150,147 @@ public void urlChangesForDifferentRequests() {
+        assertLogEvent(InternalServerException.class.getName());
     }
 
-    public String getLogEvent(String expectedMsg) {
+    @Test
+    public void getTraitSingleUsesFilter() {
+
+        String phenoUrl = "http://getTraitSingle" + UUID.randomUUID().toString();
+        ProgramBrAPIEndpoints programBrAPIEndpoints = getBrAPIEndpoints(null, phenoUrl, null);
+
+        reset(programService);
+        when(programService.getBrapiEndpoints(any(UUID.class))).thenReturn(programBrAPIEndpoints);
+        when(programService.exists(any(UUID.class))).thenReturn(true);
+
+        checkBrAPIClientExecution(phenoUrl);
+
+        Flowable<HttpResponse<String>> call = client.exchange(
+                GET("/programs/" + validProgram.getId() + "/traits/" + validVariable.getId())
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .cookie(new NettyCookie("phylo-token", "test-registered-user")), String.class
+        );
+
+        // Expect error because brapi results will not be returned
+        HttpClientResponseException e = Assertions.assertThrows(HttpClientResponseException.class, () -> {
+            HttpResponse<String> response = call.blockingFirst();
+        });
+
+        assertEquals(HttpStatus.INTERNAL_SERVER_ERROR, e.getStatus(), "Response status is incorrect");
 
-        List<ILoggingEvent> loggingEvents = loggingEventListAppender.list.stream().filter(loggingEvent -> {
-            return loggingEvent.getMessage().equals(expectedMsg);
-        }).collect(Collectors.toList());
-        ILoggingEvent apiErrorEvent = loggingEvents.get(0);
-        return apiErrorEvent.getThrowableProxy().getMessage();
+        assertLogEvent(InternalServerException.class.getName());
+    }
+
+    public void assertLogEvent(String failureClass) {
+        // Our other assertions are thrown inside the micronaut code. We need to catch them in the logs
+        String assertionClassName = org.opentest4j.AssertionFailedError.class.getName();
+        for (ILoggingEvent loggingEvent: loggingEventListAppender.list){
+            assertEquals(failureClass, loggingEvent.getThrowableProxy().getClassName(), "Was not expected failure reason");
+        }
     }
 
     public ProgramBrAPIEndpoints getBrAPIEndpoints(String coreUrl, String phenoUrl, String genoUrl){
         return ProgramBrAPIEndpoints.builder()
-                .coreUrl(Optional.of("http://" + coreUrl))
-                .genoUrl(Optional.of("http://" + genoUrl))
-                .phenoUrl(Optional.of("http://" + phenoUrl))
+                .coreUrl(coreUrl == null ? Optional.empty() : Optional.of(coreUrl))
+                .genoUrl(genoUrl == null ? Optional.empty() : Optional.of(genoUrl))
+                .phenoUrl(phenoUrl == null ? Optional.empty() : Optional.of(phenoUrl))
                 .build();
     }
+
+    public void checkBrAPIClientExecution(String expectedUrl) throws AssertionFailedError {
```

Comment:
Just had a thought on this (and if you'd rather create another story for this, I understand).  What about using a "Promise"-like approach (in Java they're called "Future"s).  Essentially something like this:

In this method, you'd create a `Future`, typed to return a `Boolean`.  In the `Answer` object you've created, have it complete the `Future` instead of using the JUnit `assertEquals`.  Lastly, have this method return the created future.  In the test method, you'd have a variable for the future, and after you initiate the API call, you'd have an `assertEquals` method that would have an expected value of `true`, and an actual value of `future.get()`.  Here's an example:

```java
public Future<Boolean> checkBrAPIClientExecution(String expectedUrl) throws AssertionFailedError {

    CompletableFuture<Boolean> future = new CompletableFuture<>();
    
    // Takes advantage of brapi library code to mimic no return results from api.
    Answer<Optional<Object>> checkBrAPIExecution = new Answer<Optional<Object>>() {
        @Override
        public Optional<Object> answer(InvocationOnMock invocation) throws AssertionFailedError {
            BrAPIClient executingBrAPIClient = (BrAPIClient) invocation.getMock();
            // Check that our url is correct
            future.complete(expectedUrl.equals(executingBrAPIClient.brapiURI()));
            return Optional.empty();
        }
    };

    ...

    return future;
}
```

and in your test method:

```java
Future<Boolean> brapiUrlCheck = checkBrAPIClientExecution(phenoUrl);

Flowable<HttpResponse<String>> call = client.exchange(
    GET("/programs/" + validProgram.getId() + "/traits?full=true")
        .contentType(MediaType.APPLICATION_JSON)
        .cookie(new NettyCookie("phylo-token", "test-registered-user")), String.class
);

...
assertEquals(true, brapiUrlCheck.get(), "Expected url was not used");
```

I think if you went this way, you wouldn't need to look at the logs at all.

Reply:
Great idea! I implemented this and it worked as expected. The changes are pushed. Thanks for the idea!